### DISCUSSION TITLE
extension extension, ggtext helper

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#11] [Ideas]
I wrote a helper function to find plot fill colors and put html color tag in title if the categories that fill represents are mentioned there.  https://github.com/EvaMaeRey/ggtextExtra  

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/e59363d3-3adf-43bc-9a12-ae6c8481a21f)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/500eb0a6-432e-4c20-b09c-ff293a5263f0)

Feedback welcome!

a couple of known issues 

- turn off fill guide as default upon applying 
- find-replace for adding hmtl tags is a little scary (male/female banana/anana problem)
- Allow missing category?
- Make html color tagging more customizable, i.e. provide default, but allow that to change.  
- probably 'strong' (bold) should be part of the default html replacement tag?  I think this is pretty much what everyone is doing.  Right now I use markdown asterisks to get the boldness.

**Comment by teunbrand** [comment]
Well well well, great minds think alike ^_^
I also had also build something similar, but as a guide.

``` r
library(gguidance)
#> Loading required package: ggplot2
library(palmerpenguins)

ggplot(penguins) +
  aes(x = flipper_length_mm,
      y = body_mass_g,
      colour = species) +
  geom_point() +
  guides(colour = guide_subtitle(
    "The {.1} penguin has the smallest average flipper length
and the {.3} have largest length
while {.2} species is in-between."
  ))
#> Warning: Removed 2 rows containing missing values or values outside the scale range
#> (`geom_point()`).
```

![](https://i.imgur.com/lz3Cfmc.png)

<sup>Created on 2024-05-03 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

But it is a major pain in the behind to do it without {ggtext} with all the manual text shaping.
Eventually I hope {marquee} with classes can become a more permanent solution.

**Reply by EvaMaeRey** [reply]
Ah!!  Amazing gguidance!   I like that you just keep building, instead of exiting the build.  

plot +
guides(colour = guide_subtitle(
    "The {.1} penguin has the smallest average flipper length
and the {.3} have largest length
while {.2} species is in-between."
  ))

Sounds like gguidance+marquee and we'll be living the dream!   (and maybe slightly more human-readable guide_subtitle input?)

**Reply by teunbrand** [reply]
You can also do `"The {.1 Adelie} penguin"`, so it needn't be that obscure. The `{.1}` is shorthand for "I can't be bothered to spell this out, please insert label here as-is".

**Comment by EvaMaeRey** [comment]
Extremely human readable.  I saw the documentation now too. Looks like there are several options for where to write names there's name flexibility (like lowercase or uppercase which my function currently does not allow).  That is good.  I think there's a little merit to 'the just type the category names and have the system recognize them and just handle the colors' approach.  I'm not sure.

**Comment by teunbrand** [comment]
Alright I have now implemented it with {marquee} and it is wonderful.
See gist for code: https://gist.github.com/teunbrand/a8237534f305ac22913a92025895b327

``` r
gist <- "https://gist.githubusercontent.com/teunbrand/a8237534f305ac22913a92025895b327/raw/fa3df47ca7c0fa89ef9120d26076359b7822322b/guide_marquee.R"
source(gist)

# A standard plot
base <- ggplot(mpg, aes(displ, hwy)) +
  geom_point()

# Inserting glyphs
base + aes(shape = drv) +
  scale_shape_discrete(
    "Cars with four wheel <<4>>, forward <<f>> or reverse <<r>> drive.",
    guide = "marquee"
  )
```

![](https://i.imgur.com/bC5dYOB.png)

``` r

# Recolouring text
base + aes(colour = drv) +
  scale_colour_discrete(
    # Same as using {.1 four wheel}, {.2 forward} and {.3 reverse}
    "Cars with {.4 four wheel}, {.f forward} or {.r reverse} drive.",
    guide = "marquee"
  )
```

![](https://i.imgur.com/7kIuTzI.png)

``` r

# Recolouring background
base + aes(colour = drv) +
  scale_colour_discrete(
    # Same as using {.1 four wheel}, {.2 forward} and {.3 reverse}
    "Cars with {.4 four wheel}, {.f forward} or {.r reverse} drive.",
    guide = guide_marquee(label_style = style_background())
  )
```

![](https://i.imgur.com/O5EaXqb.png)

<sup>Created on 2024-06-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
Whoa. The future is now!   will guide_marquee find a home in gguidance?

The poor resolution of plots was a little distracting to me, especially w/ the shapes mapping...  seems like reprex/github should allow for higher resolution when you want to share something really awesome in data viz...

**Reply by teunbrand** [reply]
Graphics device support wasn't totally consistent in these cases so I couldn't prefix the reprexes with my usual
```r
#+ setup, include=FALSE
knitr::opts_chunk$set(dev = "ragg_png")

#+ main
...rest of script...
```

**Comment by EvaMaeRey** [comment]
I like the autodetect option! .. I think the examples could maybe have the argument named for where the title goes.  I guess most people are not familiar with the argument (I think it is name?) and adjust via labs.  

#' # Automatic label detection
#' base + aes(colour = class) +
#'   scale_colour_discrete(
#'     name = "Cars including suv and minivan vehicles",
#'     guide = guide_marquee(detect = TRUE)
#'   )
#'
#' # With automatic detection every occurrance counts, which is not always good
#' base + aes(colour = drv) +
#'   scale_colour_discrete(
#'     name =  "The riffraff eats frankfurthers 4 ever",
#'     guide = guide_marquee(detect = TRUE)
#'   )

**Reply by teunbrand** [reply]
Yeah I took some effort for 3.5.0 to make sure `name` is the first argument to almost all scales, but `labs()` would work too. Also `title` is the first argument for a guide, so that could also be a good place. There are too many places to declare these :')

**Comment by EvaMaeRey** [comment]
relevant new blogpost: https://nrennie.rbind.io/blog/coloured-text-legend-ggplot-ggtext-marquee/  A ggtext v marquee plot title as color/fill decoder comparison by @nrennie.
